core/jvm/src/main/scala/scalaz/zio/lockfree/RingBuffer.scala

```diff
@@ -0,0 +1,147 @@
+    val aHead     = head
+    val aTail     = tail
+
+    var curPosition = aHead.get()
+    var curHead     = curPosition // aTail.get() => write a test!
+
+    while (curPosition < curHead + aCapacity) {
+      val idx = posToIdx(curPosition, aMask)
+      val el  = aBuf.get(idx)
+
+      if (el.isEmpty) {
+        if (el.position == curPosition) {
+          if (aBuf.compareAndSet(idx, el, BufferElement.buildValue(curPosition, a))) {
+            val curTail = aTail.get()
+            if (curTail != curPosition) updateTail(aTail, curPosition + 1) // tail.compareAndSet(curTail, curPosition + 1)
+            return true
+          } else {
+            curPosition += 1
+          }
+        } else if (el.position > curPosition) {
+          curPosition = Math.max(aTail.get(), curHead)
+        } else {
+          sys.error(s"Offer error: impossible case $el")
+        }
+      } else { // notEmpty
+        curPosition += 1
+      }
+
+      curHead = aHead.get()
+    }
+
+    false
+  }
+
+  override def poll(): Option[A] = {
+    val aCapacity = capacity
+    val aMask     = idxMask
+    val aBuf      = buf
+    val aHead     = head
+
+    var curPosition = aHead.get()
+    var spin        = true
+
+    while (spin) {
+      val nextEmptyPosition = curPosition + aCapacity
+
+      val idx = posToIdx(curPosition, aMask)
+      val el  = aBuf.get(idx)
+
+      if (!el.isEmpty) { // there's a value
+        if (el.position == curPosition) {
+          if (aBuf.compareAndSet(idx, el, BufferElement.buildEmpty(nextEmptyPosition))) {
+            updateHead(aHead, curPosition + 1)
+            return Some(el.getValue[A])
+          } else {
+            curPosition += 1
+          }
+        } else if (el.position == nextEmptyPosition) {
+          curPosition += 1
+        } else if (el.position > nextEmptyPosition) {
+          curPosition = aHead.get()
+        } else if (el.position > curPosition) { // we hit the tail
+          spin = false
+//          val curTail = tail.get()
+//          if (curTail < curPosition) updateTail(tail, curPosition) //tail.compareAndSet(curTail, curPosition)
+        } else {
+          sys.error(s"Poll error 1: impossible case ${el.position} - $curPosition")
+        }
+      } else { // empty
+        if (el.position == nextEmptyPosition) {
+          curPosition += 1
+        } else if (el.position > nextEmptyPosition) {
+          curPosition = aHead.get()
+        } else if (el.position >= curPosition) { // we hit the tail
+          spin = false
+//          val curTail = tail.get()
+//          if (curTail < curPosition) updateTail(tail, curPosition) //tail.compareAndSet(curTail, curPosition)
+        } else {
+          sys.error(s"Poll error 2: impossible case ${el.position} - $curPosition")
+        }
+      }
+    }
+
+    None
+  }
+
+  override def isEmpty(): Boolean = tail.get() == head.get()
+
+  override def isFull(): Boolean = tail.get() == head.get() + capacity - 1
+
+  @inline
```

Comment:
You can delete all these `@inline` annotations, unfortunately they will not be respected by the compiler. Also if they are private, we'll get nearly as good performance without inlining.